README.md
[![npm version](https://badge.fury.io/js/homebridge-nature-remo-sensor.svg)](https://badge.fury.io/js/homebridge-nature-remo-sensor)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](LICENSE)

# homebridge-nature-remo-sensor

Nature Remo plugin for homebridge: https://github.com/nfarina/homebridge

## feature

This is homebridge plugin to get NatureRemo sensor values with Siri and Homekit (Home app on iOS).

![temperature](./docs/temperature.gif)
![humidity](./docs/humidity.gif)


This plugin uses the "Nature Remo Cloud API" to get the temperature, humidity and illumination of NatureRemo.

Therefore, you need OAuth2 access token, which can be obtained from [home.nature.global](https://home.nature.global/).

＃ "Nature Remo Local API" doesn't have API to get temperature, humidity and illumination in v1.0.0, so this plugin uses "Nature Remo Cloud API".
＃ Also, the Home app doesn't seem to be able to create automation triggered by temperature, humidity, or illumination sensors (confirmed in iOS11).
＃ You can do this by using a third party app such as the Elgato Eve app (unconfirmed).


## Installation

```shell
npm install homebridge-nature-remo-sensor -g
```

## Configuration

### 1. Get Nature-Remo OAuth2 Access Token

Issue at [home.nature.global](https://home.nature.global/)

### 2. Set & Get Nature-Remo DeivceNickName from Remo app

Remo app → Setting → Remo → Name

![nickname1](./docs/nickname1.jpg)
![nickname2](./docs/nickname2.jpg)



### 3. Create the config.json file
```shell
$ vim ~/.homebridge/config.json
```

```js
{
  "bridge": {
    "name": "Homebridge",
    "username": "CC:22:3D:E3:CE:30",
    "port": 51826,
    "pin": "031-45-154"
  },
  "description": "Nature Remo Control",
  "accessories": [{
    "accessory": "remo-sensor",
    "name": "sensor",
    "deviceName": "<DeivceNickName>",
    "schedule": "*/5 * * * *",
    "cache": true,
    "accessToken": "<Get your access token at https://home.nature.global/>",
    "mini": false,
    "sensors": {
      "temperature": true,
      "humidity": true,
      "light": false
    }
  }]
}
```

Note:
* `schedule` is cron syntax
* `cache` If true, sensor value requests via Homebridge will return the value refreshed by the schedule (default false)
* `mini` must be `true` for Nature Remo mini
* `sensors` is optional


CHANGELOG.md
# Changelog


## [Version 1.5.0](https://github.com/hirosan18/homebridge-nature-remo-sensor/compare/v1.4.0...v1.5.0) (2022-02-19)

#### Changes

- Add Cache Option
- Change http module to axios
- Update dependencies

## [Version 1.4.0](https://github.com/hirosan18/homebridge-nature-remo-sensor/compare/v1.3.0...v1.4.0) (2021-02-25)

#### Changes

- Translate documents into English
- Support for settings in Homebridge Config UI X
- Update dependencies

## [Version 1.3.0](https://github.com/hirosan18/homebridge-nature-remo-sensor/compare/v1.2.1...v1.3.0) (2019-02-11)

#### Changes

- Change http module
- Update dependencies

## [Version 1.2.1](https://github.com/hirosan18/homebridge-nature-remo-sensor/compare/v1.2.0...v1.2.1) (2019-02-07)

#### Changes

- Add error log output

## [Version 1.2.0](https://github.com/hirosan18/homebridge-nature-remo-sensor/compare/v1.1.0...v1.2.0) (2019-02-02)

#### Changes

- Add disable option

## [Version 1.1.0](https://github.com/hirosan18/homebridge-nature-remo-sensor/compare/v1.0.0...v1.1.0) (2018-12-03)

#### Changes

- Support for Nature Remo mini
- Update dependencies

## [Version 1.0.0](https://github.com/hirosan18/homebridge-nature-remo-sensor/compare/v0.1.0...v1.0.0) (2018-07-17)

#### Changes

- Add support for light sensor


## Version 0.1.0 (2018-04-04)

#### Changes

No change log


index.js
const axios = require('axios')
const { CronJob } = require('cron')

const DEFAULT_REQUEST_PARAMS = {
  url: 'https://api.nature.global/1/devices',
  method: 'GET'
}

const TIMEOUT = 2500
const REGEX_TIMEOUT_ERROR_CODE = /E(?:(?:SOCKET)?TIMEDOUT|CONNABORTED)/

let version
let Service
let Characteristic

module.exports = homebridge => {
  version = homebridge.version
  Service = homebridge.hap.Service
  Characteristic = homebridge.hap.Characteristic

  homebridge.registerAccessory('homebridge-nature-remo-sensor', 'remo-sensor', NatureRemoSensor, true)
}

class NatureRemoSensor {
  constructor (log, config, api) {
    log('homebridge API version: ' + version)
    log('NatureRemo Init')
    this.log = log
    this.config = config
    this.name = config.name
    this.mini = config.mini ?? false
    this.deviceName = config.deviceName
    this.accessToken = config.accessToken
    this.schedule = config.schedule || '*/5 * * * *'
    this.cache = config.cache ?? false

    this.previousSensorValue = null

    const sensors = config.sensors ?? {}
    const isEnabledTemperature = sensors.temperature !== false
    const isEnabledHumidity = this.mini !== true && sensors.humidity !== false
    const isEnabledLight = this.mini !== true && sensors.light !== false

    if (this.mini) {
      log('Humidity and light sensors are disabled in NatureRemo mini')
    }

    this.informationService = new Service.AccessoryInformation()
    this.temperatureSensorService = isEnabledTemperature ? new Service.TemperatureSensor(config.name) : null
    this.humiditySensorService = isEnabledHumidity ? new Service.HumiditySensor(config.name) : null
    this.lightSensorService = isEnabledLight ? new Service.LightSensor(config.name) : null

    this.job = new CronJob({
      cronTime: this.schedule,
      onTick: () => {
        this.log('> [Schedule]')
        this.request().then((data) => {
          this.previousSensorValue = this.parseResponseData(data)
          const { humidity, temperature, light } = this.previousSensorValue
          if (this.temperatureSensorService) {
            this.log(`>>> [Update] temperature => ${temperature}`)
            this.temperatureSensorService.getCharacteristic(Characteristic.CurrentTemperature).updateValue(temperature)
          }
          if (this.humiditySensorService) {
            this.log(`>>> [Update] humidity => ${humidity}`)
            this.humiditySensorService.getCharacteristic(Characteristic.CurrentRelativeHumidity).updateValue(humidity)
          }
          if (this.lightSensorService) {
            this.log(`>>> [Update] light => ${light}`)
            this.lightSensorService.getCharacteristic(Characteristic.CurrentAmbientLightLevel).updateValue(light)
          }
          this.log('> [Schedule] finish')
        }).catch((error) => {
          this.log(`>>> [Error] "${error}"`)
          this.previousSensorValue = null
          if (this.temperatureSensorService) {
            this.temperatureSensorService.getCharacteristic(Characteristic.CurrentTemperature).updateValue(error)
          }
          if (this.humiditySensorService) {
            this.humiditySensorService.getCharacteristic(Characteristic.CurrentRelativeHumidity).updateValue(error)
          }
          if (this.lightSensorService) {
            this.lightSensorService.getCharacteristic(Characteristic.CurrentAmbientLightLevel).updateValue(error)
          }
          this.log('> [Schedule] finish')
        })
      },
      runOnInit: true
    })
    this.job.start()

    this.getTemperature = this.createGetSensorFunc('temperature')
    this.getHumidity = this.createGetSensorFunc('humidity')
    this.getLight = this.createGetSensorFunc('light')
  }

  request (option) {
    if (!this.runningPromise) {
      const options = Object.assign({}, DEFAULT_REQUEST_PARAMS, {
        headers: {
          authorization: `Bearer ${this.accessToken}`
        }
      }, typeof option === 'object' ? option : {})

      this.log('>> [request] start')
      this.runningPromise = axios(options)
        .then((response) => {
          const limit = response.headers?.['x-rate-limit-limit'] ?? 0
          const remaining = response.headers?.['x-rate-limit-remaining'] ?? 0
          this.log(`>>> [response] status: ${response.status}, limit: ${remaining}/${limit}`)
          delete this.runningPromise
          return response.data
        })
        .catch((error) => {
          const response = error?.response
          const limit = response?.headers?.['x-rate-limit-limit'] ?? 0
          const remaining = response?.headers?.['x-rate-limit-remaining'] ?? 0
          this.log(`>>> [response] status: ${response?.status ?? 'NONE'}, limit: ${remaining}/${limit}`)
          delete this.runningPromise
          throw error
        })
    }
    return this.runningPromise
  }

  parseResponseData (responseData) {
    let humidity = null
    let temperature = null
    let light = null

    let data

    if (this.deviceName) {
      data = (responseData || []).find((device, i) => {
        return device.name === this.deviceName
      })
    }
    data = data ?? (responseData || [])[0]

    if (data && data.newest_events) {
      if (data.newest_events.hu) {
        humidity = data.newest_events.hu.val
      }
      if (data.newest_events.te) {
        temperature = data.newest_events.te.val
      }
      if (data.newest_events.il) {
        light = data.newest_events.il.val
      }
    }
    return { humidity, temperature, light }
  }

  createGetSensorFunc (type) {
    return (callback) => {
      this.log(`> [Getting] ${type}`)
      const previousSensorValue = this.previousSensorValue?.[type]
      if (this.cache && typeof previousSensorValue === 'number') {
        this.log(`>>> [Getting] ${type} => ${previousSensorValue} (from cache)`)
        callback(null, previousSensorValue)
      } else {
        this.request({ timeout: TIMEOUT }).then((data) => {
          const value = this.parseResponseData(data)?.[type]
          this.log(`>>> [Getting] ${type} => ${value}`)
          callback(null, value)
        }).catch((error) => {
          this.log(`>>> [Error] "${error}"`)
          if (REGEX_TIMEOUT_ERROR_CODE.test(error.code) && typeof previousSensorValue === 'number') {
            callback(null, previousSensorValue)
          } else {
            callback(error)
          }
        })
      }
    }
  }

  getServices () {
    this.log(`start homebridge Server ${this.name}`)

    this.informationService
      .setCharacteristic(Characteristic.Manufacturer, 'Nature')
      .setCharacteristic(Characteristic.Model, 'Remo')
      .setCharacteristic(Characteristic.SerialNumber, '031-45-154')

    const services = [this.informationService]

    if (this.temperatureSensorService) {
      this.temperatureSensorService
        .getCharacteristic(Characteristic.CurrentTemperature)
        .on('get', this.getTemperature.bind(this))

      services.push(this.temperatureSensorService)
    }

    if (this.humiditySensorService) {
      this.humiditySensorService
        .getCharacteristic(Characteristic.CurrentRelativeHumidity)
        .on('get', this.getHumidity.bind(this))

      services.push(this.humiditySensorService)
    }

    if (this.lightSensorService) {
      this.lightSensorService
        .getCharacteristic(Characteristic.CurrentAmbientLightLevel)
        .on('get', this.getLight.bind(this))

      services.push(this.lightSensorService)
    }

    return services
  }
}
